### PR TITLE
Set up enhanced context A/B test

### PR DESCRIPTION
**Pull request by janhartman**
https://linear.app/sourcegraph/project/run-test-to-measure-the-impact-of-symf-and-embeddings-on-retention-28b7a712987b/overview

I created 3 new feature flags: The first one tells us if we're in the test and the second and third tell us in which group we're in. I think this setup will be fairly easy to manage - we set the first one to the % of users we want to capture and just use the second and third to achieve a 25-25-25-25 split by setting them to 50% each.
Also added a telemetry event but it might not be emitted in the best location (maybe it should happen on startup?)

## Test plan
Test locally with overrides.

### CODE REVIEW

File: vscode/src/chat/chat-view/context.ts [hunk 1]
```diff
@@ -183,6 +186,45 @@ async function getEnhancedContextFromRanker({
...
+        variant = { group: 1, strategy: 'blended' }
+    } else if (useEmbeddings) {
+        variant = { group: 2, strategy: 'embeddings' }
+    } else if (useSymf) {
+        variant = { group: 3, strategy: 'keyword' }
+    } else {
+        variant = { group: 0, strategy: 'none' }
+    }
+
+    telemetryRecorder.recordEvent('cody.embeddings.holdoutTest', 'enrolled', {
```
[the remark this answers is not in the document]
Comment (janhartman) [reply vscode/src/chat/chat-view/context.ts:220]: @hitesh-1997 I followed the instructions [here](https://linear.app/sourcegraph/project/run-test-to-measure-the-impact-of-symf-and-embeddings-on-retention-28b7a712987b#comment-c48331b1) (see "Client side telemetry events"). Good point on how it's not necessary.
@chenkc805 could we remove the telemetry and follow Hitesh's approach?

File: vscode/src/chat/chat-view/context.ts [hunk 2]
```diff
@@ -183,6 +186,45 @@ async function getEnhancedContextFromRanker({
...
+        metadata: {
+            controlVariant: variant.group,
```
[the remark this answers is not in the document]
Comment (janhartman) [reply vscode/src/chat/chat-view/context.ts:222]: Will rename if we decide to continue with the telemetry approach.

File: vscode/src/chat/chat-view/context.ts [hunk 3]
```diff
@@ -183,6 +186,45 @@ async function getEnhancedContextFromRanker({
     })
 }
 
+export async function getContextStrategy(
+    defaultStrategy: ConfigurationUseContext
+): Promise<ConfigurationUseContext> {
+    // Only run experiment if we're in VS Code
+    if (vscode.workspace.getConfiguration().get<boolean>('cody.advanced.agent.running', false)) {
```
[the remark this answers is not in the document]
Comment (janhartman) [reply vscode/src/chat/chat-view/context.ts:193]: I'd prefer to keep all relevant logic in one place, eventually this function could decide the strategy based on heuristics or something else.

File: vscode/src/chat/chat-view/context.ts [hunk 4]
```diff
@@ -183,6 +186,45 @@ async function getEnhancedContextFromRanker({
...
+    const isEnhancedContextExperiment = await featureFlagProvider.evaluateFeatureFlag(
+        FeatureFlag.CodyEnhancedContextExperiment
+    )
+    if (!isEnhancedContextExperiment) {
+        return defaultStrategy
+    }
+
+    const useEmbeddings = await featureFlagProvider.evaluateFeatureFlag(
+        FeatureFlag.CodyEnhancedContexUseEmbeddings
+    )
```
Comment (rafax) [inline vscode/src/chat/chat-view/context.ts:206]: Should we use `Promise.all` to evaluate those concurrently?

Comment (janhartman) [reply vscode/src/chat/chat-view/context.ts]: Good point, adding.

File: vscode/src/chat/chat-view/context.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (chenkc805) [reply vscode/src/chat/chat-view/context.ts:218]: Sounds good to me

**Review by chenkc805 — APPROVED** [review]

File: vscode/src/chat/chat-view/context.ts [hunk 5]
```diff
@@ -183,6 +185,36 @@ async function getEnhancedContextFromRanker({
...
+    if (useEmbeddings && useSymf) {
+        return 'blended'
+    }
+    if (useEmbeddings) {
+        return 'embeddings'
+    }
+    if (useSymf) {
+        return 'keyword'
+    }
+    return 'none'
```
Comment (hitesh-1997) [inline vscode/src/chat/chat-view/context.ts:215]: should we return `defaultStrategy` if none of the variants is satisfied

**Review by hitesh-1997 — APPROVED** [review]
Changes mostly look good to me @janhartman :)

File: vscode/src/chat/chat-view/context.ts (hunk 5 above)
Comment (janhartman) [reply vscode/src/chat/chat-view/context.ts:215]: It has to be 'none' if we're not using any of the context sources.

**Pull request merged**